### PR TITLE
feat(docs): add first explain blog post

### PR DESCRIPTION
**Pull request by bziobrowski**

### CODE REVIEW

File: blog/2023-01-26-introducing-explain.mdx [hunk 1]
```diff
@@ -0,0 +1,443 @@
...
+
+Now, we've got the right plan but don't know the reason why the database chooses
+it over the former one. To understand the difference we've to take a look at the
+documentation of the
+[now() function](https://questdb.io/docs/reference/function/date-time/#now):
+
+> Unlike sysdatetime(), it does not change within the query execution timeframe
+> and should be used in WHERE clause to filter designated timestamp column
+> relative to current time
+
```
Comment (amyshwang) [inline blog/2023-01-26-introducing-explain.mdx:433]: Could we not copy from the Docs as things might change. Maybe rephrase this?

File: blog/2023-01-26-introducing-explain.mdx [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Interval forward scan on: trades                                      |
+| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;intervals: [static=[1673862703000001,9223372036854775807] |
+
+Instead of a full scan we've only an interval scan limited to
+['2023-01-16T09:51:43.000000Z', max timestamp] time range.
+
+Now, we've got the right plan but don't know the reason why the database chooses
+it over the former one. To understand the difference we've to take a look at the
+documentation of the
+[now() function](https://questdb.io/docs/reference/function/date-time/#now):
```
Comment (amyshwang) [inline blog/2023-01-26-introducing-explain.mdx:428]:
```suggestion
it over the former one. To understand the difference, we need to understand the `sysdatetime()` function: `sysdatetime()` changes within the query execution timeframe.
```

I am guess from the context that the first query uses sysdatetime() which is costly. The function now() is not really mentioned above so using it could be confusing...

Comment (bziobrowski) [reply blog/2023-01-26-introducing-explain.mdx:428]: I agree . Maybe it'd be better to add the warning to systimestamp() and sysdate() functions and then use excerpt from the former here ?

File: blog/2023-01-26-introducing-explain.mdx [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+SELECT count(*) FROM trades WHERE price = 2.1f;
+```
+
+Query B:
+
+```questdb-sql
+SELECT count(*) FROM trades WHERE price = 2.1;
+```
+
+For hot data response times are consistent:
```
Comment (amyshwang) [inline blog/2023-01-26-introducing-explain.mdx:151]: I think this is just for my own understanding: why these are consistent? Query A tends to take longer than B in a consistent way?

File: blog/2023-01-26-introducing-explain.mdx [hunk 4]
```diff
@@ -0,0 +1,439 @@
...
+>
+  Photo by <a href="https://unsplash.com/@nguyendhn">Nguyen Dang Hoang Nhu</a>{" "}
+  via on
+  <a href="https://unsplash.com">Unsplash</a>
+</Banner>
+
+It could be a number of things, but why guess when you can check:
+
+```questdb-sql
+EXPLAIN SELECT count(*) FROM trades WHERE price = 0.0f;
```
Comment (amyshwang) [inline blog/2023-01-26-introducing-explain.mdx:172]: Query A is not 0.0f but 2.1f, does this matter?
```suggestion
EXPLAIN SELECT count(*) FROM trades WHERE price = 2.1f;
```

File: blog/2023-01-26-introducing-explain.mdx [hunk 5]
```diff
@@ -0,0 +1,439 @@
...
+
+1. Forward scan all the partitons of the table.
+1. Forward scan all rows in each partition.
+1. Apply filters to each row.
+1. Count all rows that pass it.
+
+Next, we check the plan for the second query :
+
+```questdb-sql
+EXPLAIN SELECT count(*) FROM trades WHERE price = 0.0;
```
Comment (amyshwang) [inline blog/2023-01-26-introducing-explain.mdx:204]: Again if we refer to query B why is it 0.0 here?
```suggestion
EXPLAIN SELECT count(*) FROM trades WHERE price = 2.1;
```

File: blog/2023-01-26-introducing-explain.mdx [hunk 6]
```diff
@@ -0,0 +1,439 @@
...
+| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;DataFrame                                             |
+| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Row forward scan              |
+| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Frame forward scan on: trades |
+
+Can you spot the difference?
+
+It's about the **_JIT_**.
+
+The second query is faster because it uses
+[Just-In-Time-compiled](/blog/2022/01/12/jit-sql-compiler/) filter!
```
Comment (amyshwang) [inline blog/2023-01-26-introducing-explain.mdx:221]:
```suggestion
[Just-In-Time-compiled](/docs/concept/jit-compiler/) filter!
```
I think using Doc is better than a blog post?

File: blog/2023-01-26-introducing-explain.mdx [hunk 7]
```diff
@@ -0,0 +1,439 @@
...
+FROM pos
+WHERE substring(id,1,21) = 'NSTTQFPKISYEGDLPJFEKP'
+678
+```
+
+Response time - 1.2s.
+
+A bit slow for just a few hundred records.
+
+Let's inspect the plan:
```
Comment (amyshwang) [inline blog/2023-01-26-introducing-explain.mdx:263]: Should we add the explain query here?

File: blog/2023-01-26-introducing-explain.mdx [hunk 8]
```diff
@@ -0,0 +1,439 @@
...
+| Count (1)                                                                                          |
+| &nbsp;&nbsp;&nbsp;&nbsp;Async Filter (2)                                                           |
+| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;filter: substring(id,1,21)='NSTTQFPKISYEGDLPJFEKPS'            |
+| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;workers: 14                                                    |
+| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;DataFrame (3)                                      |
+| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Row forward scan           |
+| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Frame forward scan on: pos |
+
+Query counts (1) records filtered by asynchronous, multi-threaded filter (2),
+which scans the whole table (3).
```
Comment (amyshwang) [inline blog/2023-01-26-introducing-explain.mdx:276]: the ref for (3) is unclear to me because you did not mention DataFrame...

File: blog/2023-01-26-introducing-explain.mdx [hunk 9]
```diff
@@ -0,0 +1,439 @@
...
+letters only and is 22-characters long, we can rewrite the expression with a
+`IN[to add link]` clause :
+
+```questdb-sql
+SELECT count(*)
+FROM pos
+WHERE id IN ('NSTTQFPKISYEGDLPJFEKPA','NSTTQFPKISYEGDLPJFEKPB' ... 'NSTTQFPKISYEGDLPJFEKPZ');
+```
+
+In effect, the response time shrinks to 3ms, about 400 times.
```
Comment (amyshwang) [inline blog/2023-01-26-introducing-explain.mdx:299]: 400 times less?

**Review by amyshwang — CHANGES_REQUESTED** [review]

File: blog/2023-01-26-introducing-explain.mdx (hunk 7 above)
Comment (bziobrowski) [reply blog/2023-01-26-introducing-explain.mdx:263]: It'd simply repeat the query above with explain prefix, so a bit verbose and not really offer much considering that there's a number of exaplain examples earlier  .  What do you think ?

File: blog/2023-01-26-introducing-explain.mdx (hunk 3 above)
Comment (bziobrowski) [reply blog/2023-01-26-introducing-explain.mdx]: Removed the word.  I meant to say that times a 'stable' .

File: blog/2023-01-26-introducing-explain.mdx (hunk 4 above)
Comment (bziobrowski) [reply blog/2023-01-26-introducing-explain.mdx:172]: Fixed.

File: blog/2023-01-26-introducing-explain.mdx (hunk 5 above)
Comment (bziobrowski) [reply blog/2023-01-26-introducing-explain.mdx:204]: Fixed.

File: blog/2023-01-26-introducing-explain.mdx (hunk 8 above)
Comment (bziobrowski) [reply blog/2023-01-26-introducing-explain.mdx:276]: I added (3) to both subnodes of DataFrame .

File: blog/2023-01-26-introducing-explain.mdx (hunk 9 above)
Comment (bziobrowski) [reply blog/2023-01-26-introducing-explain.mdx]: Fixed.

File: blog/2023-01-26-introducing-explain.mdx (hunk 6 above)
Comment (bziobrowski) [reply blog/2023-01-26-introducing-explain.mdx]: Quality-wise blog post is better but doc does refer to it so I think it's fine and applied your suggestion .

File: blog/2023-01-26-introducing-explain.mdx [hunk 10]
```diff
@@ -0,0 +1,439 @@
...
+## `EXPLAIN` explained
+
+[EXPLAIN](https://questdb.io/docs/reference/sql/explain) SQL keyword shows the
+execution plan (the outcome of step 3. above), which is the logical algorithm
+used to implement a given statement and thus 'explains' its performance. Even
+though it's not part of the SQL standard, many databases, e.g. PostgreSQL, MySQL or
+Oracle, do implement their own variant of `EXPLAIN`. Contrary to SQL, which is a
+high-level declarative language, `EXPLAIN` uses, often low-level, terms related
+to database implementation, for example, hash joins or index scan. Learning
+these [concepts](https://questdb.io/docs/concept/storage-model/#read-model) is
```
[the remark this answers is not in the document]
Comment (amyshwang) [reply blog/2023-01-26-introducing-explain.mdx]: I am linking it to docs/introduction/concepts/!

File: blog/2023-01-26-introducing-explain.mdx [hunk 11]
```diff
@@ -0,0 +1,439 @@
...
+| Count (1)                                                                                              |
+| &nbsp;&nbsp;&nbsp;&nbsp;Async Filter (2)                                                               |
+| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;filter: substring(id,1,21)='NSTTQFPKISYEGDLPJFEKPS'                |
+| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;workers: 14                                                        |
+| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;DataFrame (3)                                          |
+| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Row forward scan (3)           |
+| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Frame forward scan on: pos (3) |
+
+Query counts (1) records filtered by asynchronous, multi-threaded filter (2),
+which scans the whole table (3).
```
Comment (amyshwang) [inline blog/2023-01-26-introducing-explain.mdx:276]:
```suggestion
The query counts (1) records filtered by an asynchronous, multi-threaded filter (2),
which scans the whole table (3).
```

So now I understand what the numbers are referring to. The current sentence does not convert a sequence. It describes a nesting relationship, but I wonder if the number corresponds to the sequence of the operation? Does it mean that it follows the following steps:
1. record count
2. apply filter
3. scan table

I don't think so and I think the steps are in reverse. Should we use a number list to describe the steps and number the them differently? As:
1. A full table scan is triggered.
2. An asynchronous, multi-threaded filter is applied to the record.
3. Filtered records are counted.
And we number the plan accordingly too...

Maybe I get it al wrong...

Comment (bziobrowski) [reply blog/2023-01-26-introducing-explain.mdx:287]: These numbers were meant to point to pieces of execution plan, not show the sequence of operation .
You're right about the actual steps.

File: blog/2023-01-26-introducing-explain.mdx [hunk 12]
```diff
@@ -283,11 +287,13 @@ Let's used `EXPLAIN` on the query and inspect the plan:
...
 
-Query counts (1) records filtered by asynchronous, multi-threaded filter (2),
-which scans the whole table (3).
+I have annotated the plan with some numbers to illustrate the information. From
+the table, we can see that query counts (1 in the table) records filtered by an
+asynchronous, multi-threaded filter (marked as 2), which scans the whole table
+(marked as 3).
 
-Considering the number of records (678 vs 120M in the table) it seems strange
-that the index on `id` wasn't used.
```
Comment (bziobrowski) [inline blog/2023-01-26-introducing-explain.mdx]: v.s. -> vs. ?

File: blog/2023-01-26-introducing-explain.mdx [hunk 13]
```diff
@@ -0,0 +1,464 @@
...
+| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;DataFrame (3)                                          |
+| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Row forward scan (3)           |
+| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Frame forward scan on: pos (3) |
+
+I have annotated the plan with some numbers to illustrate the information. From
+the table, we can see that query counts (1 in the table) records filtered by an
+asynchronous, multi-threaded filter (marked as 2), which scans the whole table
+(marked as 3).
+
+Considering the number of records (678 rows in the result v.s. 120M in the whole
```
Comment (amyshwang) [inline blog/2023-01-26-introducing-explain.mdx:295]:
```suggestion
Considering the number of records (678 rows in the result vs 120M in the whole
```

File: blog/2023-01-26-introducing-explain.mdx [hunk 14]
```diff
@@ -0,0 +1,464 @@
...
+- How to order joins - e.g. 5 tables can be joined in 5! = 120 different orders.
+- How to optimize expressions - e.g. evaluate constant expression once instead
+  of for every row.
+
+The above is just the tip of the iceberg, a lot is happening under the hood and
+many factors influence the end performance of a query.
+
+## `EXPLAIN` explained
+
+The [EXPLAIN](https://questdb.io/docs/reference/sql/explain) SQL keyword shows
```
Comment (bluestreak01) [inline blog/2023-01-26-introducing-explain.mdx:112]:
```suggestion
The [EXPLAIN](https://questdb.io/docs/reference/sql/explain/) SQL keyword shows
```

**Review by bluestreak01 — DISMISSED** [review]

**Review by amyshwang — APPROVED** [review]
